switch prefer_deferred_runtime_asserts_over_guards in export

avikchaudhuri opened the pull request:
Summary:
In preparation for checking shape guards in export, this PR effectively switches `prefer_deferred_runtime_asserts_over_guards` to `False`, matching Dynamo.

Actually that's a lie: we switch it to `allow_complex_guards_as_runtime_asserts`, which is `False` by default but can be controlled via an internally API to be `True`. This makes the two flags synchronized, so we should be able to kill `allow_complex_guards_as_runtime_asserts` at this point.

Test Plan:
updated tests

Rollback Plan:

Differential Revision: D79734206


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

tugsbayasgalan commented on torch/export/__init__.py:
```diff
@@ -168,6 +168,7 @@ def export(
     dynamic_shapes: Optional[Union[dict[str, Any], tuple[Any, ...], list[Any]]] = None,
     strict: bool = False,
     preserve_module_call_signature: tuple[str, ...] = (),
+    prefer_deferred_runtime_asserts_over_guards: bool = False,
```
I feel you also need to propagate it to export_for_training as well.

avikchaudhuri replied:
@angelayi Yes, intentional: I pushed it up to the top level to act as a switch to manage risk.

@tugsbayasgalan The tests didn't need me to push this to `export_for_training`. Since it's on the way to deprecation, leave as is?

tugsbayasgalan replied:
Hmm interesting. Today torch.export.export calls into torch.export._trace._export_for_training at pre-dispatch level.

avikchaudhuri replied:
To be clear, `_export_for_training` does have this flag. Anyway, I can add it.

avikchaudhuri committed:
switch prefer_deferred_runtime_asserts_over_guards in export (#160111)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/160111

In preparation for checking shape guards in export, this PR effectively switches `prefer_deferred_runtime_asserts_over_guards` to `False`, matching Dynamo.

Actually that's a lie: we switch it to `allow_complex_guards_as_runtime_asserts`, which is `False` by default but can be controlled via an internally API to be `True`. This makes the two flags synchronized, so we should be able to kill `allow_complex_guards_as_runtime_asserts` at this point.

Test Plan:
updated tests

Rollback Plan:

Differential Revision: D79734206
